### PR TITLE
Parallelize Tests in CI (+ Refactor and Improvements)

### PR DESCRIPTION
**Pull request by aminya**
### Description of the Change
This:
- On Windows, runs core renderer tests using two Azure agents. This in addition to double the overall test speed, will allow rerunning the failed ones in case some fail due to timeouts.
- allows requesting any tests on any operating system
- parallelizes core renderer tests by using separate child processes for each file. 
- allows rerunning only the failed tests rather than restarting the whole CI.
- reduces Windows timeouts by running the tests on windows-2019.

### Verification
The CI passes

### Release Notes
Same as the description

### Closes
This supersedes these old pull requests: 
#18270

### CODE REVIEW

**Comment by DeeDeeG** [comment]
We could re-write this PR to not include the heavy use of Azure DevOps templates. Which would make for smaller and more targeted changes, and be easier to review.

Let us know; We'd be very glad to see this improvement land at this repo, so CI can run faster here.

**Comment by aminya** [comment]
> We could re-write this PR to not include the heavy use of Azure DevOps templates.

I don't think rewriting this is a good idea. They should merge #21081 before this. Not using templates just makes a soup of code that it is hard to maintain.

File: script/vsts/platforms/linux.yml [hunk 1]
```diff
@@ -30,12 +30,26 @@ jobs:
...
         displayName: Update npm
 
-      - task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
-        displayName: Restore node_modules cache
+      - task: Cache@2
+        displayName: Cache node_modules
         inputs:
-          keyfile: 'package.json, script/vsts/platforms/linux.yml, **/package-lock.json, !**/node_modules/**/package-lock.json, !**/.*/**/package-lock.json'
-          targetfolder: '**/node_modules, !**/node_modules/**/node_modules'
-          vstsFeed: 'bae1bc26-220d-43c7-a955-4de039370de2'
```
Comment (Duv4lC4ruso) [inline script/vsts/platforms/linux.yml]: #19441 
#10326 
#20511

Comment (aminya) [reply script/vsts/platforms/linux.yml]: Please remove the spam comment which references incorrect issues. If I report your account that is much worse @Duv4lC4ruso.

Comment (DeeDeeG) [reply script/vsts/platforms/linux.yml]: I've already reported it as spam. I don't see any particularly legitimate activity on the account. Sorry if it's a real person, but the account activity looks like a bot.

**Comment by sadick254** [comment]
@aminya I think this needs rebasing.

**Comment by aminya** [comment]
@sadick254 This is ready to go! 😉

**Review by sadick254 — COMMENTED** [review]
@aminya Thank you for your contribution. Looking at this PR 2 questions come to mind.
- Does this affect the overall speed of running `script/test`
- What is the impact of spawning a `child_process` for every spec file?

Another thing i have noted is, this PR changes the behaviour of running `script/test`. This PR makes it mandatory to specify the suite one would like to run. Initially `script/test` would run all the core tests without extra `args`.

**Comment by aminya** [comment]
> @aminya Thank you for your contribution. Looking at this PR 2 questions come to mind.
> 
> * Does this affect the overall speed of running `script/test`
> * What is the impact of spawning a `child_process` for every spec file?
> 

Thank you for asking. I haven't noticed any negative impact on performance. The speed bottleneck of running the tests is not in spawning the child_process (which might take less than 1 second), but in running the tests themselves. 

Using separate child_process allows us to selectively choose what we want to run, and rerun them if they fail (#21295). Using this I have reduced the overall performance and test pass rate.

With that being said there is still room for improvement. If you have noticed we use `async.series` which is not ideal. I have been mostly concerned with increasing the overall test experience. To maximize the speed of running these, we should apply some other changes and change this to `async.parallel`. I can do that after we merged #21295.

> Another thing i have noted is, this PR changes the behaviour of running `script/test`. This PR makes it mandatory to specify the suite one would like to run. Initially `script/test` would run all the core tests without extra `args`.

Yes, and I think this is a better approach. However, if you don't like this, it has an easy fix. Here instead of throwing an error, we can run all the tests:
https://github.com/atom/atom/pull/21109/files#diff-0e2bc0f03fe54e5e0fff129205e6ee92R294

**Comment by sadick254** [comment]
@aminya Thanks for the clarification. 
> Yes, and I think this is a better approach. However, if you don't like this, it has an easy fix. Here instead of throwing an error, we can run all the tests

No need. I think this is ok. 👍

**Pull request merged**